### PR TITLE
support justifyLastLine

### PR DESCRIPTION
**Pull request by YoheiZuho**
This is:

- [ ] a bugfix
- [x] a new feature
- [ ] refactoring
- [ ] additional unit tests

Checklist:

- [ ] Changes are covered by unit tests
  - [x] Changes are covered by existing unit tests
  - [ ] New unit tests have been added
- [x] Code style is respected
- [x] Commit message explains **why** the change is made (see https://github.com/erlang/otp/wiki/Writing-good-commit-messages)
- [x] CHANGELOG.md contains a short summary of the change and a link to the pull request if applicable
- [x] Documentation is updated as necessary

### Why this change is needed?
justifyLastLine” is a flag used for the ‘Justify Distributed’ option, which is enabled only when the value of ‘Horizontal’ in the EXCEL setting ‘FormatCells’ is ‘Distributed’. When the option is enabled, an indent of a certain width is added before and after the text in the cell.
This option is often used in Japanese-speaking countries.

### CODE REVIEW

**Comment by oleibman** [comment]
You will need to fix the Phpstan problems and the unit tests that broke as a result of your change. Then you need to make sure that the new property in Alignment is used when calculating the alignment hash code. Then you need to add some tests to demonstrate that your change is working - at a minimum, set the new property in some cell on a spreadsheet (and leave it unset in other cells), write out the spreadsheet, read it back in, and confirm that the property is set or not as desired.

**Comment by YoheiZuho** [comment]
I'd fix the Phpstan problems and the unit tests.

**Comment by oleibman** [comment]
Good work on correcting the earlier problems. You still need to include the new field when generating the hash code for Alignment, and to add unit test(s).

**Comment by YoheiZuho** [comment]
> You still need to include the new field when generating the hash code for Alignment, and to add unit test(s).

Excuse me, but which file should I add the test to?

File: src/PhpSpreadsheet/Reader/Xlsx/Styles.php [hunk 1]
```diff
@@ -265,6 +265,8 @@ public function readAlignmentStyle(Alignment $alignment, SimpleXMLElement $align
         if ($horizontal !== '') {
             $alignment->setHorizontal($horizontal);
         }
+        $justifyLastLine = (string) $this->getAttribute($alignmentXml, 'justifyLastLine');
```
Comment (oleibman) [inline src/PhpSpreadsheet/Reader/Xlsx/Styles.php:268]: This isn't quite right. Because justifyLastLine can be null as well as true/false, if justifyLastLine isn't in the Xml, you should leave it as its default property. Your code will set it to false. It probably doesn't make a difference to the end result, but we strive to be "correct".

File: tests/PhpSpreadsheetTests/Writer/Xlsx/Issue3443Test.php [hunk 2]
```diff
@@ -37,6 +37,7 @@ public function testNonDefaultAlign(): void
         $rsheet = $reloadedSpreadsheet->getActiveSheet();
         $expected1 = [
             'horizontal' => 'center',
+            'justifyLastLine' => false,
```
Comment (oleibman) [inline tests/PhpSpreadsheetTests/Writer/Xlsx/Issue3443Test.php:40]: This will probably be affected by the change I suggested above to Reader/Xlsx/Styles.

File: tests/PhpSpreadsheetTests/Writer/Xlsx/Issue3443Test.php [hunk 3]
```diff
@@ -78,6 +79,7 @@ public function testDefaultAlign(): void
         $rsheet = $reloadedSpreadsheet->getActiveSheet();
         $expected1 = [
             'horizontal' => 'general',
+            'justifyLastLine' => false,
```
Comment (oleibman) [inline tests/PhpSpreadsheetTests/Writer/Xlsx/Issue3443Test.php:82]: This will probably be affected by the change I suggested above to Reader/Xlsx/Styles.

File: tests/PhpSpreadsheetTests/Style/AlignmentTest.php [hunk 4]
```diff
@@ -84,6 +84,17 @@ public function testHorizontal(): void
         self::assertEquals(0, $cell3->getStyle()->getAlignment()->getIndent());
     }
 
+    public function testJustifyLastLine(): void
+    {
+        $this->spreadsheet = new Spreadsheet();
+        $sheet = $this->spreadsheet->getActiveSheet();
+        $cell1 = $sheet->getCell('A1');
+        $cell1->setValue('ABC');
+        $cell1->getStyle()->getAlignment()->setHorizontal(Alignment::HORIZONTAL_DISTRIBUTED);
```
Comment (oleibman) [inline tests/PhpSpreadsheetTests/Style/AlignmentTest.php:93]: Your test is not doing enough (or you want to add an additional test in Reader/Xlsx). We want to write and then read the spreadsheet to prove that both Write and Read handle the new attribute correctly. Look for examples of this by searching for "writeAndReload" in the PhpSpreadsheetTests Reader/Xlsx directory.

**Review by oleibman — CHANGES_REQUESTED** [review]

File: src/PhpSpreadsheet/Style/Alignment.php [hunk 5]
```diff
@@ -475,6 +512,7 @@ public function getHashCode(): string
 
         return md5(
             $this->horizontal
+            . $this->justifyLastLine
```
Comment (oleibman) [inline src/PhpSpreadsheet/Style/Alignment.php:515]: Another case where the 3-pronged value will cause problems. Probably change this to:
```php
. (($this->justifyLastLine === null) ? 'null' : ($this->justifyLastLine ? 't' : 'f'))
```

**Comment by oleibman** [comment]
Despite the number of changes I've requested, you're in pretty good shape. I expect this will be ready to go soon.

File: src/PhpSpreadsheet/Reader/Xlsx/Styles.php (hunk 1 above)
Comment (oleibman) [reply src/PhpSpreadsheet/Reader/Xlsx/Styles.php:268]: resolved In latest push.

File: tests/PhpSpreadsheetTests/Style/AlignmentTest.php (hunk 4 above)
Comment (oleibman) [reply tests/PhpSpreadsheetTests/Style/AlignmentTest.php:93]: Resolved in latest push.

**Comment by oleibman** [comment]
Thank you for your contribution.
